File: hidden-api/src/main/java/android/app/ActivityTaskManager.java
package android.app;

import android.os.Build;

import androidx.annotation.RequiresApi;

import java.util.List;

/**
 * Compile-time stub for the @hide ActivityTaskManager.
 * Doesn't need to use @RefineAs, because ActivityTaskManager is not visible in public SDK.
 */
public class ActivityTaskManager {
    // since android 10
    @RequiresApi(Build.VERSION_CODES.Q)
    public static IActivityTaskManager getService() {
        throw new RuntimeException("Stub!");
    }
    // since android 12
    @RequiresApi(Build.VERSION_CODES.S)
    public static ActivityTaskManager getInstance() {
        throw new RuntimeException("Stub!");
    }
    @RequiresApi(Build.VERSION_CODES.S)
    public List<ActivityManager.RunningTaskInfo> getTasks(int maxNum) {
        throw new RuntimeException("Stub!");
    }
}

File: hidden-api-contract/src/main/java/com/xaxaxax/moonclicker/hiddenapi/contracts/ActivityTaskManager.kt
package com.xaxaxax.moonclicker.hiddenapi.contracts

import android.os.Build

/**
 * Split out of ActivityManager in Android 10; the class does not exist on API 27–28.
 */
internal val ACTIVITY_TASK_MANAGER = StubContracts(
    stubs = mapOf("android.app.ActivityTaskManager" to "android.app.ActivityTaskManager"),
    contracts = listOf(
        MemberContract(
            owner = "android.app.ActivityTaskManager",
            member = MethodMember(
                name = "getService",
                returns = "android.app.IActivityTaskManager",
                static = true,
            ),
            sinceApi = Build.VERSION_CODES.Q,
        ),
        // Unused by production today; present and matching only from API 31.
        MemberContract(
            owner = "android.app.ActivityTaskManager",
            member = MethodMember(
                name = "getInstance",
                returns = "android.app.ActivityTaskManager",
                static = true,
            ),
            sinceApi = Build.VERSION_CODES.S,
        ),
        MemberContract(
            owner = "android.app.ActivityTaskManager",
            member = MethodMember(
                name = "getTasks",
                parameters = listOf("int"),
                returns = "java.util.List",
            ),
            sinceApi = Build.VERSION_CODES.S,
        ),
    ),
)
